Soft matrix support

azedinez opened the pull request:
Added support for non Affymetrix platforms via download of SOFT matrices directly from GEO. These matrices are already quantile normalised, but to be comparable to the output of JUSTRMA, a log2 transformation is applied (if applicable). 

Tested with added profile test_soft. Also tested on a couple of beadchip mouse datasets. 

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/differentialabundance/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/differentialabundance _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

pinin4fjords commented on CHANGELOG.md:
```diff
@@ -3,6 +3,10 @@
 
+## v1.4dev - 2023-05-15
+
+- [] - Added support for non-Affymetrix arrays via automatic download of SOFT matrices in GEO
```
This will need to go in the 1.3.0dev below

pinin4fjords commented on nextflow_schema.json:
```diff
@@ -24,7 +24,7 @@
                     "description": "A string identifying the technology used to produce the data",
                     "help_text": "Currently 'rnaseq' or 'affy_array' may be specified.",
-                    "enum": ["rnaseq", "affy_array"],
+                    "enum": ["rnaseq", "affy_array", "non_affy_array"],
```
I'm not convinced this is the most helpful renaming. e.g. we may at some point have a module that deals with illumina-style arrays. Should this not be something that indicates the pulling of soft files from GEO?

azedine-healx replied:
Agreed, sorry for the lack of inspiration :) I replaced with geo_soft_file. Let me know if that's any better

pinin4fjords replied:
looks good

pinin4fjords commented on workflows/differentialabundance.nf:
```diff
@@ -22,16 +22,24 @@ if (params.study_type == 'affy_array'){
+  // then a GSE study identifier must be provided
+} else if (params.study_type == 'non_affy_array' && params.querygse != "" && params.features_metadata_cols != ""){
+    if (params.querygse) {
+        ch_querygse = Channel.of([exp_meta, params.querygse, params.features_metadata_cols])
```
You don't need to place the param in a channel- it should be used directly from params when needed.

azedine-healx replied:
Makes sense. I've removed it as input and put directly as external parameter. Sorry, still getting my head around how Nextflow works :)

pinin4fjords reviewed (changes requested):
I'm going to come back with more review soon- a couple of comments to start.

Also, you're going to have to appease eclint in the R code- see the CI errors.

pinin4fjords commented on workflows/differentialabundance.nf:
```diff
@@ -214,7 +232,12 @@ workflow DIFFERENTIALABUNDANCE {
+    }
+    else if (params.study_type == 'geo_soft_file') {
+        ch_norm = VALIDATOR.out.assays
+        ch_matrix_for_differential = ch_norm
```
suggest changing the logic to set `ch_matrix_for_differential = ch_norm` for all non-rnaseq.

pinin4fjords commented on workflows/differentialabundance.nf:
```diff
@@ -359,20 +383,31 @@ workflow DIFFERENTIALABUNDANCE {
-
-    ch_contrast_variables
-        .combine(ch_all_matrices.map{ it.tail() })
+    if(params.study_type != "geo_soft_file") {
```
You don't need to duplicate this whole section. You can just do like:

```
if(params.study_type != "geo_soft_file") {
   ch_mat = ch_raw.combine(ch_processed_matrices)
}else{
   ch_mat = ch_processed_matrices
}
```

... and then use `ch_mat` in the original chain of ops

azedinez replied:
Makes a lot more sense... thanks and sorry for the bad code!

pinin4fjords reviewed (commented):
Sorry, another batch of comments ;-)

In general, try to maintain consistency with existing behaviours and avoid duplications where possible.

azedine-healx commented on nextflow_schema.json:
```diff
@@ -81,6 +81,13 @@
+                "querygse": {
+                    "type": "string",
+                    "default": "None",
+                    "description": "Alternative to CEL archive: the GSE identifier as found in GEO",
```
Can I contest this one? There's no mention of Affymetrix, just an alternative to CELs which can also apply to Affymetrix?

pinin4fjords replied:
Imagine we add other modalities such as the special non-CEL array formats from Illumina. We'd need to remember to come back and fix this. It's easier if you're just specific about what you mean from the start.

pinin4fjords replied:
(and AFAIK CEL === affy)

azedinez replied:
I meant user may want to use the GSE identifier for an Affy study and use SOFT instead of CEL in that case as well. Although I get your point, and amended it to "Use SOFT files from GEO by providing the GSE study identifier"

azedine-healx commented on workflows/differentialabundance.nf:
```diff
@@ -137,49 +147,57 @@ workflow DIFFERENTIALABUNDANCE {
+        ch_generic_array_input = ch_input
+            .join(ch_querygse)
+
+        READ_FROM_SOFT(ch_generic_array_input)
```
Done. I've also added it for the two JUSTRMA modules

azedine-healx commented on modules/local/read_from_soft.nf:
```diff
@@ -0,0 +1,24 @@
+process READ_FROM_SOFT {
```
Makes perfect sense. Will do, but in the interest of reducing cognitive load for myself, can we first merge this PR (to dev) with this as a local module first? I can then commence the PR to submit READ_FROM_SFOT as an nf-core module, and later open a PR in diffab to reflect the changes. Would that be ok?

pinin4fjords replied:
Sorry, I don't want to create precedents for anyone else adding modules- plus these things have a way of never actually happening ;-).

I'll PR the module for you soon if you like, give me a few days.

azedinez replied:
Actually let me handle it. I'd like to contribute to the open source efforts a bit more, and this seems like a good way to do it. Will try and do that in the next couple of weeks

azedinez commented:
Apart from moving READ_FROM_SOFT to nf-core, I believe I've addressed all the comments so far. Let me know if there's anything else you'd like to me look at

pinin4fjords commented:
@azedinez you'll see above that that central module was done a few weeks ago. All that need to be done now is the workflow installation and integration.

azedinez commented:
Hi @pinin4fjords. Thanks for adding the geoquery as a module. Sorry that I haven't worked on this recently, I haven't given up on it. I will get that done in due course!

pinin4fjords commented:
> Hi @pinin4fjords. Thanks for adding the geoquery as a module. Sorry that I haven't worked on this recently, I haven't given up on it. I will get that done in due course!

No worries, just didn't want you to have missed the module going in :-)

azedinez commented:
@pinin4fjords Done :)

azedinez commented:
Hi @pinin4fjords, just wanted to make sure you've seen I've made the changes required. Let me know if there's anything else I need to do on this PR

pinin4fjords reviewed (changes requested):
Getting there! Sorry for the delay.

Still a few artefacts to remove from the local module. 

Also, you're currently setting `ch_in_raw` to the output of the new modules, but the `ch_norm` to the output of the validator (which doesn't normalise).

If you expect the input matrix to be normalised, then use `ch_in_norm` to collect the initial geoquery/getgeo output. If it's raw, then maybe we need to normalise it (via a LIMMA_NORM analagous to DESEQ2_NORM).

azedinez commented on workflows/differentialabundance.nf:
```diff
@@ -220,8 +247,15 @@ workflow DIFFERENTIALABUNDANCE {
         ch_norm = ch_validated_assays.normalised
+    }
+    else if (params.study_type == 'geo_soft_file') {
+        ch_norm = VALIDATOR.out.assays
```
Yes, the output of the SOFT matrix is normalised

azedinez commented on docs/usage.md:
```diff
@@ -57,6 +57,20 @@ This is a numeric square matrix file, comma or tab-separated, with a column for
+
+`--study_type geo_soft_file` and `--querygse [GSE study ID]`
+
+enables the pipeline to download normalised SOFT matrices automatically (note that even though Affymetrix arrays are also supported in the SOFT matrix track, it is recommended to work from CEL files in this case). Importantly, user must provide a valid set of fields for the features metadata relevant for the platforms e.g. for GPL570 (Affymetrix Plus 2.0 arrays) this could be done with
```
Yes but the default in the pipeline doesn't match the columns in the SOFT so might crash the run downstream. Also, each array platform has their own meta cols so it's never the same that are returned, so we'd need to set up correct meta cols defaults for virtually all array platforms in GEO.

azedinez commented:
Apologies for delay in addressing this. I believe I've addressed all your comments but let me know if I need to address more stuff. Otherwise feel free to merge.

pinin4fjords commented:
@azedinez I had to make a number of fixes and tweaks to make this actually work (and the tests are now running in CI). Could you please test locally to make sure outputs are as you would like?

azedinez commented:
@pinin4fjords Tested locally and runs. All output looks as expected. I have just added the soft config to the pipeline for convenience.

pinin4fjords reviewed (approved):
Great- we've got there. Thanks for the work @azedinez !
